apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Apis.Util;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Core;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// The SessionPool is the preferred method of managing Spanner Sessions.
+    /// It performs pooling of sessions with an evict timer along with configurable settings to control the maximum
+    /// number of pooled and active sessions.
+    /// To create a session using the sessionpool, use the extension method on SpannerClient.
+    ///   var session = await spannerClient.CreateSessionFromPoolAsync(project, spannerInstance, database, cancellationtoken);
+    /// To release a session back to the sessionpool:
+    ///   session.ReleaseToPool();
+    /// If you use the SessionPool, you must make sure sessions are properly released back into the pool and are not simply GC'd.
+    /// Allowing a session to GC will incur penalties on the current process as the session will count towards the maximum allowed
+    /// and it will incur a penalty on other spanner processes because it will be an hour before the server frees the session if 
+    /// its not properly deleted.
+    /// </summary>
+    public static class SessionPool
+    {
+        private static readonly TimeSpan s_shutDownTimer = TimeSpan.FromSeconds(60);
+
+        // This member holds information we'll use when the session later gets released.
+        private static readonly ConcurrentDictionary<Session, SessionPoolKey>
+            s_sessionsInUse = new ConcurrentDictionary<Session, SessionPoolKey>();
+
+        private static int s_sessionsCreating;
+        private static readonly ConcurrentQueue<TaskCompletionSource<int>> s_waitQueue = new ConcurrentQueue<TaskCompletionSource<int>>();
+        private static readonly object s_waitSync = new object();
+
+        private static ConcurrentDictionary<SessionPoolKey, SessionPoolImpl> GlobalPool { get; } =
+            new ConcurrentDictionary<SessionPoolKey, SessionPoolImpl>();
```

Comment:
Possibly rename this to something like `sessionPoolsByDatabase`? That would make it clearer (IMO) that there's two levels of pooling going on.